lib/bolt_ext/puppetdb_inventory.rb

```diff
@@ -0,0 +1,191 @@
+
+      def initialize(options)
+        @settings = load_config(File.expand_path('~/.puppetlabs/client-tools/puppetdb.conf'))
+        @settings.merge!(options)
+        validate
+      end
+
+      def load_config(filename)
+        if File.exist?(filename)
+          JSON.parse(File.read(filename)).fetch('puppetdb', {})
+        # Only try to use the default token if it exists, so we don't override
+        # cert/key with a non-existent token
+        elsif File.exist?(DEFAULT_TOKEN)
+          { 'token' => DEFAULT_TOKEN }
+        else
+          {}
+        end
+      end
+
+      def [](key)
+        @settings[key]
+      end
+
+      def validate
+        unless @settings['server_urls']
+          raise "server_urls must be specified in the config file or with --url"
+        end
+        unless @settings['cacert']
+          raise "cacert must be specified in the config file or with --cacert"
+        end
+        if !@settings['token'] && !@settings['cert'] && !@settings['key']
+          raise "token or cert/key must be provided"
+        end
+        if (@settings['cert'] && !@settings['key']) ||
+           (!@settings['cert'] && @settings['key'])
+          raise "cert and key must be specified together"
+        end
+      end
+    end
+
+    class CLI
+      def initialize(args)
+        @args = args
+        @cli_opts = {}
+        @parser = build_parser
+      end
+
+      def build_parser
+        parser = OptionParser.new('') do |opts|
+          opts.on('--cacert CACERT', "Path to the CA certificate") do |cacert|
+            @cli_opts['cacert'] = cacert
+          end
+          opts.on('--cert CERT', "Path to the certificate") do |cert|
+            @cli_opts['cert'] = cert
+          end
+          opts.on('--key KEY', "Path to the private key") do |key|
+            @cli_opts['key'] = key
+          end
+          opts.on('--token-file TOKEN', "Path to the token file") do |token|
+            @cli_opts['token'] = token
+          end
+          opts.on('--url URL', "The URL of the PuppetDB server to connect to") do |url|
+            @cli_opts['server_urls'] = [url]
+          end
+          opts.on('--file FILE', "Path to the inventory file to use") do |file|
+            @inventory_file = File.expand_path(file)
+          end
+          opts.on('-h', '--help', "Display help") do |_|
+            @show_help = true
+          end
+        end
+        parser.banner = <<-BANNER
+  Usage: bolt-inventory-pdb --file <inventory-file> [--url <url>] [auth-options]
+
+  Populate the nodes in an inventory file based on PuppetDB queries.
+
+        BANNER
+        parser
+      end
+
+      def run
+        @parser.permute(@args)
+
+        if @show_help
+          puts @parser.help
+          return 0
+        end
+
+        unless @inventory_file
+          raise "--file is a required option"
+        end
+
+        config = Config.new(@cli_opts)
+        @puppetdb_client = Client.new(config)
```

Comment:
This call should be `Client.from_config`. I also found it useful to add a `--trace` option that enables printing `e.backtrace` in the rescue at line 175 for debugging.

Reply:
My patches to get it working
```
diff --git a/lib/bolt_ext/puppetdb_inventory.rb b/lib/bolt_ext/puppetdb_inventory.rb
index 66ce9e1..29bb86c 100755
--- a/lib/bolt_ext/puppetdb_inventory.rb
+++ b/lib/bolt_ext/puppetdb_inventory.rb
@@ -69,7 +69,8 @@ module Bolt
       DEFAULT_TOKEN = File.expand_path('~/.puppetlabs/token')

       def initialize(options)
-        @settings = load_config(File.expand_path('~/.puppetlabs/client-tools/puppetdb.conf'))
+        @settings =
+          default_token(load_config(File.expand_path('~/.puppetlabs/client-tools/puppetdb.conf')))
         @settings.merge!(options)
         validate
       end
@@ -77,12 +78,18 @@ module Bolt
       def load_config(filename)
         if File.exist?(filename)
           JSON.parse(File.read(filename)).fetch('puppetdb', {})
+        else
+          {}
+        end
+      end
+
+      def default_token(conf)
         # Only try to use the default token if it exists, so we don't override
         # cert/key with a non-existent token
-        elsif File.exist?(DEFAULT_TOKEN)
-          { 'token' => DEFAULT_TOKEN }
+        if File.exist?(DEFAULT_TOKEN)
+          { 'token' => DEFAULT_TOKEN }.merge(conf)
         else
-          {}
+          conf
         end
       end

@@ -134,6 +141,9 @@ module Bolt
           opts.on('--file FILE', "Path to the inventory file to use") do |file|
             @inventory_file = File.expand_path(file)
           end
+          opts.on('--trace', "Enable stack trace on errors") do |_|
+            @trace = true
+          end
           opts.on('-h', '--help', "Display help") do |_|
             @show_help = true
           end
@@ -160,7 +170,7 @@ module Bolt
         end

         config = Config.new(@cli_opts)
-        @puppetdb_client = Client.new(config)
+        @puppetdb_client = Client.from_config(config)

         unless File.readable?(@inventory_file)
           raise "Can't read the inventory file #{@inventory_file}"
@@ -174,6 +184,7 @@ module Bolt
         return 0
       rescue StandardError => e
         puts "Error: #{e}"
+        puts e.backtrace if e.backtrace && @trace
         return 1
       end
```

Reply:
Doh, I refactored some of that for testability, but I guess I didn't actually update all the uses of it in the code. Unit tests only get so far. :P